service/header/sync.go

```diff
@@ -213,83 +251,106 @@ func (s *Syncer) processIncoming(ctx context.Context, maybeHead *ExtendedHeader)
 	return pubsub.ValidationAccept
 }
 
-// TODO(@Wondertan): Number of headers that can be requested at once. Either make this configurable or,
-//  find a proper rationale for constant.
-var requestSize uint64 = 512
-
 // syncTo requests headers from locally stored head up to the new head.
-func (s *Syncer) syncTo(ctx context.Context, newHead *ExtendedHeader) error {
+func (s *Syncer) syncTo(ctx context.Context, newHead *ExtendedHeader) {
 	head, err := s.store.Head(ctx)
 	if err != nil {
-		return err
+		log.Errorw("getting head during sync", "err", err)
+		return
 	}
 
 	if head.Height == newHead.Height {
-		return nil
+		return
 	}
 
-	log.Infow("syncing headers", "from", head.Height, "to", newHead.Height)
-	defer log.Info("synced headers")
+	log.Infow("syncing headers",
+		"from", head.Height,
+		"to", newHead.Height)
+	err = s.doSync(ctx, head, newHead)
+	if err != nil {
+		log.Errorw("syncing headers",
+			"from", head.Height,
+			"to", newHead.Height,
+			"err", err)
+		return
+	}
 
-	start, end := uint64(head.Height)+1, uint64(newHead.Height)
-	for start <= end {
-		amount := end - start + 1
-		if amount > requestSize {
-			amount = requestSize
-		}
+	log.Infow("finished syncing",
+		"from", head.Height,
+		"to", newHead.Height,
+		"elapsed time", s.state.End.Sub(s.state.Start))
+}
 
-		headers, err := s.getHeaders(ctx, start, amount)
-		if err != nil && len(headers) == 0 {
-			return err
+// doSync performs actual syncing updating the internal SyncState
+func (s *Syncer) doSync(ctx context.Context, oldHead, newHead *ExtendedHeader) (err error) {
+	from, to := uint64(oldHead.Height)+1, uint64(newHead.Height)
+
+	s.stateLk.Lock()
+	s.state.ID++
+	s.state.FromHeight = from
+	s.state.ToHeight = to
+	s.state.FromHash = oldHead.Hash()
+	s.state.ToHash = newHead.Hash()
+	s.state.Start = time.Now()
+	s.stateLk.Unlock()
+
+	for processed := 0; from < to; from += uint64(processed) {
+		processed, err = s.processHeaders(ctx, from, to)
+		if err != nil && processed == 0 {
+			break
```

Comment:
I see the only way this can err out but continue looping is if there is an error appending the `found` headers to the store, in which case the loop continues and the value of `err` resets, meaning if there was an err but the loop continues successfully, the err likely gets logged in the store, but if the next loop is successful, the value of `err` gets set to nil and then is reported to syncState as nil -- can this happen?

Reply:
> Also, this can error when a node fails to request headers from Exchange.
 
Is there retry logic for this already built in? Like retry from a different peer or in case of core, retry request?

>  If some sync try failed, but the subsequent one is successful,

The problem here is adjacency. If sync requests header at height 100 but doesn't get it, why would it try to sync 101 immediately and try to apply it to store if it'll just get rejected due to not being adjacent?

We should probably implement some retry logic here or fail out. If we are requesting headers but can't apply them to store due to adjacency issues, we should fail out the node since the functionality of requesting headers is then useless no? Since every subsequent header won't get added to store?